### PR TITLE
OAK-9980: Index Purging Logic fails when trying to delete :oak:mount-…

### PR DESCRIPTION
**Pull request by tihom88**
…libs* (read-only) node under index definition for composite nodestore

### CODE REVIEW

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/IndexUtils.java [hunk 1]
```diff
@@ -147,4 +155,38 @@ public static void checkLibsIsReadOnly(Persistence p) throws RepositoryException
         }
     }
 
+    public static boolean isIndexDisabledAndHiddenNodesDeleted(NodeStore store, String path) throws CommitFailedException {
+        NodeState nodeState = NodeStateUtils.getNode(store.getRoot(), path);
+        if (!nodeState.getProperty("type").getValue(Type.STRING).equals("disabled")) {
```
Comment (thomasmueller) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/IndexUtils.java:161]: Comment: getProperty could return null, but as this is just a test, I think it's fine and we don't need to check for null. But in production code, we would need to verify. In the test case, I actually prefer shorter / simpler code, even if it's slightly dangerous. The only problem is when someone moves the code to /src/main. What about adding a comment

// could throw NPE, but we don't care as it's a test

So this won't be forgotten when the code is moved

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/Persistence.java [hunk 2]
```diff
@@ -138,15 +140,22 @@ public static Persistence openComposite(File globalDir, File libsDir, Config con
...
                 globalStore).addMount("libs", libsStore).
                 setPartialReadOnly(true).build();
-        result.repository = openRepsitory(cn, config.indexDir);
+        result.repository = openRepsitory(result.compositeNodestore, config.indexDir);
         result.session = result.repository.login(createCredentials());
         return result;
     }
 
+    public NodeStore getCompositeNodestore() throws Exception {
+        if (compositeNodestore == null){
```
Comment (thomasmueller) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/Persistence.java]: Nitpick: space before { is missing

**Review by thomasmueller — APPROVED** [review]

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/IndexUtils.java [hunk 3]
```diff
@@ -147,4 +155,38 @@ public static void checkLibsIsReadOnly(Persistence p) throws RepositoryException
         }
     }
 
+    public static boolean isIndexDisabledAndHiddenNodesDeleted(NodeStore store, String path) throws CommitFailedException {
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/IndexUtils.java]: `CommitFailedException` is not thrown here. Please remove it.

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/Persistence.java [hunk 4]
```diff
@@ -138,15 +140,22 @@ public static Persistence openComposite(File globalDir, File libsDir, Config con
...
+            throw new Exception("persistence object was not opened in composite mode");
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/composite/blueGreen/Persistence.java]: would it be better to throw an `IllegalStateException` here?

File: oak-run/src/test/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersionIT.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        purgeOldIndexVersionLogger.starting();
+        PurgeOldIndexVersion purgeOldIndexVersion = new PurgeOldIndexVersion();
+        Persistence p = Persistence.openComposite(globalDir, libsDir, config);
+        NodeStore n = p.getCompositeNodestore();
+        purgeOldIndexVersion.execute(n, true, 1, Arrays.asList("/oak:index"));
+
+        String purgeLog = "Found some index need to be purged over base'/oak:index/test':" +
+                " '[/oak:index/test-1 base=/oak:index/test versioned product=1 custom=0 operation:DELETE_HIDDEN_AND_DISABLE]'";
+        Assert.assertEquals(1, purgeOldIndexVersionLogger.getLogs().size());
+        Assert.assertEquals(purgeLog, purgeOldIndexVersionLogger.getLogs().get(0).toString());
```
Comment (fabriziofortino) [inline oak-run/src/test/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersionIT.java]: no need to call toString here
```suggestion
        Assert.assertEquals(purgeLog, purgeOldIndexVersionLogger.getLogs().get(0));
```

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

**Review by fabriziofortino — APPROVED** [review]

**Review by fabriziofortino — APPROVED** [review]

**Pull request merged**